File: AlgorithmsPackPythonSolutions/377. Combination Sum IV.py
class Solution:
    def combinationSum4(self, nums: List[int], target: int) -> int:
        MAX = 5000 + 1
        memory = [-1] * MAX
        memory[0] = 1
        def combi(target):
            if target < 0:
                return 0
            if target == 0:
                return 1
            if memory[target] != -1:
                return memory[target]
            ret = 0
            for value in nums:
                ret += combi(target - value)
            memory[target] = ret
            return ret
        return combi(target)

File: AlgorithmsPackCPP/377. Combination Sum IV.cpp
const int MAX = 5000+1;
int memory[MAX];
vector<int>nums;
class Solution {
public:
    int combi(int target){
        if(target<0)
            return 0;
        if(target==0)
            return 1;
        
        auto&ret = memory[target];
        if(ret!=-1)
            return ret;
        
        ret=0;
        for(auto&value:nums)
            ret+=combi(target-value);
        return ret;
    }
    
    int combinationSum4(vector<int>& nums_, int target) {
        nums = nums_;
        memset(memory,-1,sizeof(memory));
        return combi(target);
    }
};